core/multigrid/pgm.cpp

```diff
@@ -199,10 +210,290 @@ void Pgm<ValueType, IndexType>::generate()
+    }
+}
+#endif
+
+
+template <typename IndexType>
+void generate_non_local_map(
+    const std::vector<experimental::distributed::comm_index_type>& recv_offsets,
+    array<IndexType>& non_local_agg, array<IndexType>& non_local_col_map,
+    array<IndexType>& renumber)
+{
+    auto exec = renumber.get_executor();
+    auto non_local_size = non_local_agg.get_size();
+    array<IndexType> part_id(exec, non_local_size);
+    array<IndexType> index(exec, non_local_size);
 
-    this->set_multigrid_level(prolong_row_gather, coarse_matrix,
-                              restrict_sparsity);
+    for (int i = 0; i + 1 < recv_offsets.size(); i++) {
+        for (auto j = recv_offsets.at(i); j < recv_offsets.at(i + 1); j++) {
+            part_id.get_data()[j] = i;
+            index.get_data()[j] = j;
+        }
+    }
+    // do it in host currently.
+    auto it = detail::make_zip_iterator(
+        part_id.get_data(), non_local_agg.get_data(), index.get_data());
+    // prepare tuple <part_id, local_agg, index>
+    // sort by <part_id, local_agg> or did segment sort
+    std::sort(it, it + non_local_size);
+
+    renumber.get_data()[0] = 0;
+    // renumber (prefix_sum) with not equal <part_id, local_agg>
+    for (int i = 1; i < non_local_size; i++) {
+        if (part_id.get_const_data()[i] != part_id.get_const_data()[i - 1] ||
+            non_local_agg.get_const_data()[i] !=
+                non_local_agg.get_const_data()[i - 1]) {
+            renumber.get_data()[i] = renumber.get_data()[i - 1] + 1;
+        } else {
+            renumber.get_data()[i] = renumber.get_data()[i - 1];
+        }
+    }
+    renumber.get_data()[non_local_size] =
+        renumber.get_data()[non_local_size - 1] + 1;
+    // create col map
+    // for each thread i, col_map[tuple[i].index] = map[i]
+    for (int i = 0; i < non_local_size; i++) {
+        non_local_col_map.get_data()[index.get_data()[i]] =
+            renumber.get_data()[i];
+    }
+}
+
+template <typename IndexType>
+void compute_communication(
+    const std::vector<experimental::distributed::comm_index_type> recv_offsets,
+    const array<IndexType>& non_local_agg, const array<IndexType>& renumber,
+    std::vector<experimental::distributed::comm_index_type>& new_recv_size,
+    std::vector<experimental::distributed::comm_index_type>& new_recv_offsets,
+    array<IndexType>& new_recv_gather_idxs)
+{
+    new_recv_offsets.at(0) = 0;
+    for (int i = 0; i < new_recv_size.size(); i++) {
+        new_recv_size.at(i) =
+            renumber.get_const_data()[recv_offsets.at(i + 1)] -
+            renumber.get_const_data()[recv_offsets.at(i)];
+        new_recv_offsets.at(i + 1) =
+            new_recv_offsets.at(i) + new_recv_size.at(i);
+    }
+    IndexType non_local_num_agg = new_recv_offsets.back();
+    new_recv_gather_idxs.resize_and_reset(non_local_num_agg);
+    for (int i = 0; i < non_local_agg.get_size(); i++) {
+        new_recv_gather_idxs.get_data()[renumber.get_const_data()[i]] =
```

Comment:
I am not sure the implicit assumption that new_recv_gather_idxs is on the host is safe.